### PR TITLE
Fix runtime for: lazy, Printexc, refs, obj, hashtbl

### PR DESCRIPTION
**Pull request by anmonteiro**
Fixes the runtime for the following features / modules

- `Sys`
- refs
- `Lazy`
- OCaml classes
- `Hashtbl`
- `Printexc`

### CODE REVIEW

File: jscomp/stdlib-412/stdlib_modules/obj.ml [hunk 1]
```diff
@@ -40,7 +40,6 @@ external raw_field : t -> int -> raw_data = "caml_obj_raw_field"
 external set_raw_field : t -> int -> raw_data -> unit
                                           = "caml_obj_set_raw_field"
 
-external new_block : int -> int -> t = "caml_obj_block"
```
Comment (EduardoRFS) [inline jscomp/stdlib-412/stdlib_modules/obj.ml:43]: Do you know why this was removed?

**Review by EduardoRFS — COMMENTED** [review]
The changes seems to be okay, but I would like if we could document this somewhere, for the next update, because this is definitely gonna bite us in the future

File: jscomp/runtime/caml_sys.ml [hunk 2]
```diff
@@ -82,33 +82,40 @@ let caml_sys_system_command _cmd = 127
...
     return process.cwd()
   }|}]
 
 
+let caml_sys_executable_name () : string =
+  if Js.typeof [%raw{|process|}] = "undefined" then ""
+  else
+    let argv = [%raw{|process.argv|}] in
+    if Js.testAny argv then ""
+    else Caml_array_extern.unsafe_get argv 0
```
[the remark this answers is not in the document]
Comment (anmonteiro) [reply jscomp/runtime/caml_sys.ml:96]: I don't know if that's true for `#!/usr/bin/env node` scripts for example. and other tools might wanna implement `process` too.

File: jscomp/stdlib-412/stdlib.mli [hunk 3]
```diff
@@ -1301,13 +1301,13 @@ type 'a ref = { mutable contents : 'a }
 external ref : 'a -> 'a ref = "%makemutable"
 (** Return a fresh reference containing the given value. *)
 
-external ( ! ) : 'a ref -> 'a = "%field0"
+external ( ! ) : 'a ref -> 'a = "%bs_ref_field0"
```
[the remark this answers is not in the document]
Comment (anmonteiro) [reply jscomp/stdlib-412/stdlib.mli:1304]: correct. I missed it at first.

File: jscomp/stdlib-412/stdlib_modules/hashtbl.ml [hunk 4]
```diff
@@ -38,8 +38,7 @@ and ('a, 'b) bucketlist =
 *)
 
 let ongoing_traversal h =
-  Obj.size (Obj.repr h) < 4 (* compatibility with old hash tables *)
-  || h.initial_size < 0
+  h.initial_size < 0
```
[the remark this answers is not in the document]
Comment (anmonteiro) [reply jscomp/stdlib-412/stdlib_modules/hashtbl.ml:41]: `Obj.size` doesn't work in this scenario. And yes, but it's a mirror of the 406 stdlib too.

File: jscomp/stdlib-412/stdlib_modules/camlinternalOO.ml [hunk 5]
```diff
@@ -27,6 +33,13 @@ let copy o =
 
 (**** Compression options ****)
 (* Parameters *)
+#if BS then(* {!Translobj.oo_prim : string -> lambda} not by slot *)
+module Sys = struct
+  external word_size : unit -> int = "%word_size"
+  let word_size = word_size ()
```
[the remark this answers is not in the document]
Comment (anmonteiro) [reply jscomp/stdlib-412/stdlib_modules/camlinternalOO.ml:39]: I think so. The advantage of this piece of code, as I understand it, is that it makes the diff minimal for other functions in the namespace that call `Sys.word_size`.

**Comment by anmonteiro** [comment]
re: documenting the diff, I think it might be enough to just diff the entire stdlib when we bring in a new version. Because we're planning on upgrading fairly regularly, I don't expect that to be a problem. 

How does that sound?

File: jscomp/runtime/caml_sys.ml (hunk 2 above)
[the remark this answers is not in the document]
Comment (EduardoRFS) [reply jscomp/runtime/caml_sys.ml:96]: `#!/usr/bin/env node` will not do it but other tools will

File: jscomp/stdlib-412/stdlib_modules/camlinternalOO.ml (hunk 5 above)
[the remark this answers is not in the document]
Comment (EduardoRFS) [reply jscomp/stdlib-412/stdlib_modules/camlinternalOO.ml:39]: got it

**Comment by EduardoRFS** [comment]
@anmonteiro ok for me, but we should take notes of this somewhere

**Comment by anmonteiro** [comment]
I'll work on a CONTRIBUTING file one of these days. I'll make sure to document the upgrades.

**Review by EduardoRFS — APPROVED** [review]
Overall I think this is okay, but I'm a bit scared of the unknowns and we should take notes of them

anime girl scared

**Pull request merged**